### ISSUE TITLE
Computation of conditional attributes

### ISSUE THREAD

**Issue by erik-buchholz** [Yasoz/DiffTraj#10]
Dear Yuanshao,

I am currently reproducing your work with the goal of using DiffTraj as a baseline method in comparison with a novel approach. Therefore, we would like to run your model on our datasets to allow for a fair comparison. Could you give me some information on how the conditional information `[depature, total_dis, total_time, total_len, avg_dis, avg_speed, sid, eid]` can be computed for a new dataset? 
Unfortunately, I could not find any respective information in the repository. 
Based on the details found in the paper and this repo, I made the following assumptions:

For the numerical normalisation of attribute $attr$, you used the formula

$$
attr_{\text{norm}} = \frac{attr - \mu}{\sigma}
$$

1. Departure Time: _Categorical_ Attribute of the time of day in 5-minute steps (see https://github.com/Yasoz/DiffTraj/issues/1). _Not Normalised_
2. Total Distance: _Numerical_ Attribute normalised via mean and standard deviation (see above).
    I assume this is the sum of the distances between all two consecutive points within the trajectory? Or is this just the distance between start and end point?
3. Total Duration: _Numerical_ Attribute normalised via mean and standard deviation (see above).
    This is total duration. But in what unit? Seconds? Minutes?
4. Total Length: _Numerical_ Attribute normalised via mean and standard deviation (see above).
    Number of points in the trajectory?
5. Avg Distance: _Numerical_ Attribute normalised via mean and standard deviation (see above).
    Average distance (in m) between 2 consecutive locations?
6. Average Speed: _Numerical_ Attribute normalised via mean and standard deviation (see above).
    `Total Length / Total Duration`
7. Start Grid ID: _Categorical_ Attribute. _Not Normalised_ Values from $0 - 140$
   How did you compute the grid? I assume you used `divide_grids` from utils.py. But what did you use as the boundaries of the datasets? Just the maximal/minimal lat/lon values? What grid number did you use? How did you numerate the grid cells? From top left to bottom right? 
8. End Grid ID: _Categorical_ Attribute. _Not Normalised_
   Same questions as for (7).

Could you confirm whether these assumptions are correct and fill the missing gaps?
Moreover, could you provide details on how to compute attributes 7 and 8?

Also, if you could share any code for the computation of these values, I would greatly appreciate it.

I appreciate your help. We would love to include your approach in our publication.

Best regards,
Erik

**Comment by Yasoz** [comment]
Dear Erik,

Thanks for your interest in our work. Your comprehension and calculation of these attributes are basically in line with the design of the paper. Here are some clarifications to some of your questions:

1. Departure Time, Start Grid ID, and End Grid ID do not need to be normalized.
2. The calculation of Total distance is the cumulative distance between all two consecutive points.
3. Total Duration is in seconds, but I don't think that will matter due to the normalization we use.
4. Total length is the number of points in the trajectory. Note that since we used linear interpolation to fix the trajectory length, here is the original number of trajectory points.
5. Average distance (in m) is the distance between 2 consecutive locations.
6. Average Speed is calculated by Total Distance / Total Duration.
7.  For all normalized attributes, we use the z-score formula you pointed out. All distances should be calculated as geographic distances.

I hope the answers above are helpful and good luck with your work.

Yours,
Yuanshao

**Comment by erik-buchholz** [comment]
Dear Yuanshao,

Thanks a lot for the prompt response! 

Just one follow-up question: Could you elaborate on how you derive the values for `sid` and `eid`? 
I.e., how exactly do you determine the grid cell ids of the start and points? Do you build the grid and enumerate all cells, and then, check whether a certain location is contained within?

Best regards,
Erik

**Comment by Yasoz** [comment]
Sorry for the missing question. You can start by counting the overall trajectory's maximum/minimum latitude/longitude as a boundary value and then calculate the grid. It doesn't matter what the grid number is and in what order, just follow the same settings. In this paper, I have used both 12*12 and 16*16 grid numbers, which should be adjusted according to your data range. Once the boundaries are set, calculating the values of sid and eid is straightforward, just by determining the grid number where the start and end points are located.

**Comment by erik-buchholz** [comment]
Perfect, thank you so much! 

One last question: If I would like to run the model without the conditional guidance, i.e., `DiffTraj-wo/Con` in your paper, I should use the `Model` instead of `Guide_UNet` and set the `extra_embed` to none when calling it. Is this correct? Or is there anything else that needs to be done? In the config, I could change `'attr_dim': ` to `0` in this case?

**Comment by erik-buchholz** [comment]
So, what would the code for DiffTraj-wo/Con have to look like?

Like this?
```python
class Uncond_UNet(nn.Module):
    """
    This is the model without conditional information: DiffTraj-wo/Con
    """
    def __init__(self, config):
        super(Uncond_UNet, self).__init__()
        self.config = config
        self.unet = Model(config)

    def forward(self, x, t):
        pred_noise = self.unet(x, t, extra_embed=None)
        return pred_noise
```

Or did you still include the place embedding as it does not actually make use of the attributes?

I.e., like this:

```python
class Uncond_UNet(nn.Module):
    """
    This is the model without conditional information: DiffTraj-wo/Con
    """
    def __init__(self, config):
        super(Uncond_UNet, self).__init__()
        self.config = config
        self.ch = config.model.ch * 4
        self.attr_dim = config.model.attr_dim
        self.guidance_scale = config.model.guidance_scale
        self.unet = Model(config)

        # Remove guide embedding but keep place embedding
        # self.guide_emb = WideAndDeep(self.ch)
        self.place_emb = WideAndDeep(self.ch)

    def forward(self, x, t, attr=None):
        # Remove guide
        # guide_emb = self.guide_emb(attr)
        # Keep place embedding
        place_vector = torch.zeros((x.shape[0], 8), device=x.device)
        place_emb = self.place_emb(place_vector)

        # cond_noise needs to be removed b/c we don't have guide information
        # cond_noise = self.unet(x, t, guide_emb)
        uncond_noise = self.unet(x, t, place_emb)
        
        # Not sure how to define pred_noise
        # pred_noise = cond_noise + self.guidance_scale * (cond_noise -
                                                         uncond_noise)
        # Maybe like this?
        pred_noise = uncond_noise
        
        return pred_noise
```

Thanks a lot for your help!

**Comment by erik-buchholz** [comment]
I made my second question a separate issue as it has nothing to do with the original issue. I think this one can be closed.
